src/llama-vocab.cpp

```diff
@@ -1966,3 +1966,19 @@ int32_t llama_detokenize_impl(
 
     return total <= text_len_max ? total : -total;
 }
+
+std::string llama_detokenize(const struct llama_vocab & vocab, const std::vector<llama_token> & tokens, bool special) {
+    std::string text;
+    text.resize(std::max(text.capacity(), tokens.size()));
+    int32_t n_chars = llama_detokenize_impl(vocab, tokens.data(), (int32_t)tokens.size(), &text[0], (int32_t)text.size(), false, special);
+    if (n_chars < 0) {
+        text.resize(-n_chars);
+        n_chars = llama_detokenize_impl(vocab, tokens.data(), (int32_t)tokens.size(), &text[0], (int32_t)text.size(), false, special);
+        GGML_ASSERT(n_chars <= (int32_t)text.size());  // whitespace trimming is performed after per-token detokenization
+    }
+
+    text.resize(n_chars);
+
+    // NOTE: the original tokenizer decodes bytes after collecting the pieces.
+    return text;
+}
```

Comment:
It might be better to not introduce this function and instead maintain a string in the sampler context that is resized as needed (see the infill sampler for a similar idea). The main goal would be to avoid memory allocations that would currently happen here for each token. Can be done in follow-up PR

Reply:
In this case, the function is only used for initialization with the sequence breakers so the impact should hopefully not be too great. In the case of Nemo, this process resulst in the initial default 4 sequence breakers transforming into over 2000 sequence breakers because @pi6am’s overlap code includes any tokens that overlap with the 4 sequence breakers. But this process only happens on initialization and the processed sequence breakers are then kept in token form from then on.

Reply:
Got it - I missed that this is used only during init.